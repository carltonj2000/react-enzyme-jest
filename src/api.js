const addUser = user => console.log(user);

export default { addUser };
