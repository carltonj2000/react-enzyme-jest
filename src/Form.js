import React, { Component } from "react";
import api from "./api";

export class Form extends Component {
  state = { name: "", email: "", phone: "", optIn: true };
  handleChange = str => e => {
    this.setState({ [str]: e.currentTarget.value });
  };
  handleCkBx = e => {
    this.setState(pre => ({ optIn: !pre.optIn }));
  };
  handleSubmit = e => {
    e.preventDefault();
    return api.addUser(this.state);
  };
  render() {
    return (
      <form data-testid="addUserForm" onSubmit={this.handleSubmit}>
        <input
          data-testid="name"
          onChange={this.handleChange("name")}
          value={this.state.name}
        />
        <input
          data-testid="email"
          onChange={this.handleChange("email")}
          value={this.state.email}
        />
        <input
          data-testid="phone"
          onChange={this.handleChange("phone")}
          value={this.state.phone}
        />
        <input
          data-testid="optIn"
          type="checkbox"
          onChange={this.handleCkBx}
          checked={this.state.optIn}
        />
        <button data-testid="submitButton" onClick={this.handleSubmit}>
          Submit
        </button>
      </form>
    );
  }
}

export default Form;
