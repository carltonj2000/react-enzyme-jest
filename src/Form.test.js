import React from "react";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";

import { Form } from "./Form";
import api from "./api";

configure({ adapter: new Adapter() });

const updateInput = (wrapper, instance, newValue) => {
  const input = wrapper.find(instance);
  input.simulate("change", { currentTarget: { value: newValue } });
  return wrapper.find(instance).props().value;
};

describe("<Form />", () => {
  it("receive promotions default is true", () => {
    const wrapper = shallow(<Form />);
    const optIn = wrapper.find('[data-testid="optIn"]');
    expect(optIn.props().checked).toBe(true);
  });
  it("submits the form", () => {
    jest
      .spyOn(api, "addUser")
      .mockImplementation(() => Promise.resolve({ data: "New user added" }));
    const wrapper = shallow(<Form />);
    expect(updateInput(wrapper, '[data-testid="name"]', "Carlton")).toBe(
      "Carlton"
    );
    expect(updateInput(wrapper, '[data-testid="email"]', "cj@me.com")).toBe(
      "cj@me.com"
    );
    expect(updateInput(wrapper, '[data-testid="phone"]', "1234")).toBe("1234");
    wrapper
      .find('[data-testid="optIn"]')
      .simulate("change", { target: { checked: false } });
    expect(wrapper.find('[data-testid="optIn"]').props().checked).toBe(false);
    wrapper
      .find('[data-testid="addUserForm"]')
      .simulate("submit", { preventDefault: () => "hi" });
    expect(api.addUser).toHaveBeenCalledWith({
      name: "Carlton",
      email: "cj@me.com",
      phone: "1234",
      optIn: false
    });
  });
  it("matches snapshot", () => {
    const wrapper = shallow(<Form />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
