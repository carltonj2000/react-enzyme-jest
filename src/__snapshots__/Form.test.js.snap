// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> matches snapshot 1`] = `
<form
  data-testid="addUserForm"
  onSubmit={[Function]}
>
  <input
    data-testid="name"
    onChange={[Function]}
    value=""
  />
  <input
    data-testid="email"
    onChange={[Function]}
    value=""
  />
  <input
    data-testid="phone"
    onChange={[Function]}
    value=""
  />
  <input
    checked={true}
    data-testid="optIn"
    onChange={[Function]}
    type="checkbox"
  />
  <button
    data-testid="submitButton"
    onClick={[Function]}
  >
    Submit
  </button>
</form>
`;
